test/erlfmt_markdown_SUITE.erl

```diff
@@ -0,0 +1,175 @@
+    lists:map(
+        fun (Filename) ->
+            {ok, Content} = file:read_file(Filename),
+            check_markdown(binary_to_list(Content))
+        end,
+        Filenames
+    ).
+
+find_markdown_filenames_in(Path) ->
+    {ok, BaseFilenames} = file:list_dir_all(Path),
+    Filenames =
+        lists:map(fun (Filename) -> filename:join([Path, Filename]) end, BaseFilenames),
+    lists:filter(fun (Filename) -> filename:extension(Filename) == ".md" end, Filenames).
+
+markdown_string(Config) when is_list(Config) ->
+    S =
+        "# Heading\n"
+        "ignore for now\n"
+        "```erlang formatted key\n"
+        "hello(mike, joe, robert)\n"
+        "```\n"
+        "ignore for now\n"
+        "```erlang unformatted\n"
+        "goodbye(mike, joe, robert)\n"
+        "```\n"
+        "The previous unformatted version, should match this formatted version\n"
+        "```erlang formatted\n"
+        "goodbye(mike, joe, robert)\n"
+        "```\n"
+        "This unformatted version, should match the first formatted version with the same key\n"
+        "```erlang unformatted key\n"
+        "hello(mike, joe, robert)\n"
+        "```\n",
+    check_markdown(S).
+
+check_markdown(Content) ->
+    Sections = string:split(Content, "```", all),
+    lists:foldl(fun check_section/2, {text, maps:new()}, Sections).
+
+check_section(_Text, {text, State}) ->
+    {code, State};
+check_section(Text, {code, State}) ->
+    [FirstLine, CurrentCode] = string:split(Text, "\n"),
+    Spec = string:split(FirstLine, " ", all),
+    case Spec of
+        [Erlang | [ToFormat | Keys]] ->
+            case string:prefix(Erlang, "erl") of
+                nomatch ->
+                    {text, State};
+                _ ->
+                    Key = string:join(Keys, " "),
+                    case maps:find(Key, State) of
+                        error ->
+                            case ToFormat of
+                                "formatted" ->
+                                    check_fmt(CurrentCode, CurrentCode),
+                                    {text, maps:put(Key, CurrentCode, State)};
+                                "unformatted" ->
+                                    {text, maps:put(Key, CurrentCode, State)}
+                            end;
+                        {ok, PreviousCode} ->
+                            case ToFormat of
+                                "formatted" ->
+                                    check_fmt(PreviousCode, CurrentCode);
+                                "unformatted" ->
+                                    check_fmt(CurrentCode, PreviousCode)
+                            end,
+                            {text, maps:remove(Key, State)}
+                    end
+            end;
+        _ ->
+            {text, State}
+    end.
+
+check_fmt(Unformatted, Expected) ->
+    NewlyFormatted = format(Unformatted, 80),
+    ?assertEqual(string:trim(NewlyFormatted), string:trim(Expected)).
```

Comment:
If you go for the suggestion about changing how checking works, would also add here a check that formatting is idempotent - that the expected string after formatting is equal to itself. I think right now this only happens in the loop above if we first hit a formatted example followed by an unformatted one, in the reverse case it's not checked.

Reply:
Good catch :)